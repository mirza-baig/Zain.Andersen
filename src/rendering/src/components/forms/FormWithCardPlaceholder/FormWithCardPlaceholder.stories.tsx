// Global
import { SitecoreContext } from '@sitecore-jss/sitecore-jss-react';
import { Story, Meta } from '@storybook/react';
import { createComponentFactory } from 'lib/mocks/mock-placeholder';
// Local
import FormWithCardPlaceholder, { FormWithCardPlaceholderProps } from './FormWithCardPlaceholder';
import defaultData from './FormWithCardPlaceholder.mock-data';

const componentFactory = createComponentFactory();
export default {
  title: 'Forms/FormWithCardPlaceholder',
  component: FormWithCardPlaceholder,
  decorators: [
    (Story) => <SitecoreContext componentFactory={componentFactory}>{Story()}</SitecoreContext>,
  ],
} as Meta;
const Template: Story<FormWithCardPlaceholderProps> = (props) => (
  <FormWithCardPlaceholder {...props} />
);
export const Default = Template.bind({});
Default.args = defaultData;
