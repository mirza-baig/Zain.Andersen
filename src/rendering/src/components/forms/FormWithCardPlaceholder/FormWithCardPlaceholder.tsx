import { withDatasourceCheck, Placeholder } from '@sitecore-jss/sitecore-jss-nextjs';
import { Feature } from 'src/.generated/Feature.EnterpriseWeb.model';
import classNames from 'classnames';
import { getEnum } from 'lib/utils';
import { Component } from 'src/helpers/Component';
import { ComponentBackgroundVariants } from 'src/helpers/Component/Component';

export type FormWithCardPlaceholderProps =
  Feature.EnterpriseWeb.RenewalByAndersen.Components.Forms.FormWithCardPlaceholder.FormWithCardPlaceholder;
const FormWithCardPlaceholder = (props: FormWithCardPlaceholderProps) => {
  const { rendering, fields } = props;
  const style = getEnum<ComponentBackgroundVariants>(fields?.backgroundColor) ?? 'white';
  const isCardPresent = rendering?.placeholders?.card?.length > 0;
  const formColClasses = classNames(
    'col-span-12',
    isCardPresent
      ? ' max-md:border-b max-md:border-solid max-md:border-gray md:col-span-5'
      : ' md:col-span-4 md:col-start-5 '
  );
  return (
    <Component
      variant="full"
      gap="gap-x-0"
      padding="px-0"
      backgroundVariant={style}
      dataComponent="forms/formwithcardplaceholder"
      {...props}
    >
      <div className="col-span-12 py-l">
        <div className="grid grid-cols-2 gap-x-0 max-lg:px-m md:max-w-screen-lg md:grid-cols-12 lg:mx-auto">
          <div className={formColClasses}>
            <Placeholder name="form" rendering={rendering} />
          </div>
          {isCardPresent && (
            <div className="col-span-12  max-md:py-s md:col-span-6 md:col-start-7 md:my-2 md:border-l md:border-solid md:border-gray md:px-s [&>div]:!h-auto">
              <div className="flex">
                <Placeholder name="card" rendering={rendering} />
              </div>
            </div>
          )}
        </div>
      </div>
    </Component>
  );
};

export default withDatasourceCheck()<FormWithCardPlaceholderProps>(FormWithCardPlaceholder);
