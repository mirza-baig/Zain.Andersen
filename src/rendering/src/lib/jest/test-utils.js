// Global
import { render, fireEvent, waitFor } from '@testing-library/react';
import { ComponentPropsContext } from '@sitecore-jss/sitecore-jss-nextjs';
import { FastSitecoreContext } from 'lib/overrides/FastSitecoreContext';

// Lib
import { createSampleComponentFactory } from 'lib/mocks/mock-placeholder';

export const eeSitecoreContext = {
  route: {
    name: 'Foo',
    placeholders: {},
  },
  context: {
    pageEditing: true,
    site: {
      name: 'JssTestWeb',
    },
    language: 'en',
  },
};

export const renderComponent = (
  Component,
  {
    componentProps = {},
    staticProps = {},
    componentFactory = createSampleComponentFactory(),
    sitecoreContext = {
      route: {
        name: 'Foo',
        placeholders: {},
      },
    },
  }
) => {
  const mockLayoutData = {
    sitecore: sitecoreContext,
  };

  return render(
    <ComponentPropsContext value={staticProps}>
      <FastSitecoreContext componentFactory={componentFactory} layoutData={mockLayoutData}>
        <Component {...componentProps} />
      </FastSitecoreContext>
    </ComponentPropsContext>
  );
};

export const snapshot = (Component, mockData) => {
  const component = renderComponent(Component, mockData);
  expect(component.container.firstChild).toMatchSnapshot();
  return component;
};

export const hasDataComponent = async (component, name) => {
  await waitFor(() => {
    const nodes = component.baseElement.querySelectorAll(`[data-component="${name}"]`);
    return expect(nodes.length).toBe(1);
  });
};

export const mockDateAs = (epochMs) => {
  if (typeof epochMs !== 'number') {
    console.error(
      'mockDateAs only accepts epoch values in milliseconds to prevent timezones from factoring in to Date creation.'
    );
  }

  jest.spyOn(global.Date, 'now').mockImplementation(() => new Date(epochMs).valueOf());
};

export const testFieldError = async (Component, mockData, { label, value, error }) => {
  const component = renderComponent(Component, mockData);
  const field = component.getByLabelText(label);
  fireEvent.change(field, { target: { value } });
  field.blur();
  await waitFor(() => {
    const err = component.queryByText(error);
    return expect(err).not.toBe(null);
  });
};
