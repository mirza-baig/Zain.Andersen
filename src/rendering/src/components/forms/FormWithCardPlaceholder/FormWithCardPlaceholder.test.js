// Global
import { hasDataComponent, snapshot } from 'lib/jest/test-utils';
// Local
import FormWithCardPlaceholder from './FormWithCardPlaceholder';
import defaultData from './FormWithCardPlaceholder.mock-data';

it('renders correctly', async () => {
  const component = snapshot(FormWithCardPlaceholder, { componentProps: defaultData });
  await hasDataComponent(component, 'forms/formwithcardplaceholder');
});
