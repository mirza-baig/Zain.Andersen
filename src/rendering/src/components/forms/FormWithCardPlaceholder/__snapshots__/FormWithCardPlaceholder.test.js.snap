// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  data-component="forms/formwithcardplaceholder"
  id="id971698437"
>
  <div
    class="my-8 w-full theme-gray bg-light-gray"
  >
    <div
      class="section-grid grid grid-cols-2 md:grid-cols-12 px-0 gap-x-0 "
    >
      <div
        class="col-span-12 py-l"
      >
        <div
          class="grid grid-cols-2 gap-x-0 max-lg:px-m md:max-w-screen-lg md:grid-cols-12 lg:mx-auto"
        >
          <div
            class="col-span-12  max-md:border-b max-md:border-solid max-md:border-gray md:col-span-5"
          >
            <div
              style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
            >
              <h2>
                Form
              </h2>
              <p>
                JSS component is missing React implementation. See the developer console for more information.
              </p>
            </div>
          </div>
          <div
            class="col-span-12 max-md:py-s md:col-span-6 md:col-start-7 md:my-2 md:border-l md:border-solid md:border-gray md:px-s [&>div]:!h-auto"
          >
            <div
              class="flex"
            >
              <div
                style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
              >
                <h2>
                  ProductPreviewCard
                </h2>
                <p>
                  JSS component is missing React implementation. See the developer console for more information.
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
